File: 14/list.14.1.swift
import Foundation

// 書籍では省略されてるデータクラスなど --
struct Customer {
    struct Id {
        let id: Int
    }
    let id: Id
    let possessionPoint: PurchasePoint
    
    var isEnable: Bool {
        // 本来は何かの条件でユーザが有効か無効か決まる
        // 今回は偶数のみ有効
        id.id % 2 == 0
    }
}

struct Comic {
    struct Id {
        let id: Int
    }
    let id: Id
    let currentPurchasePoint: PurchasePoint
    
    var isEnable: Bool {
        // 本来は何かの条件でユーザが有効か無効か決まる
        // 今回は偶数のみ有効
        id.id % 2 == 0
    }
}

struct PurchasePoint {
    let amount: Int
}
// --

class PurchasePointPayment {
    let customerId: Customer.Id
    let comicId: Comic.Id
    let consumptionPoint: PurchasePoint
    let paymentDateTime: Date
    
    // パラメータを受け取ってインスタンスを生成する
    init?(customer: Customer, comic: Comic) throws {
        // [BAD]
        // if elseが多くて入れ子になっているため結局正常な処理がどこか分かりにくい
        if customer.isEnable {
            // idを登録
            self.customerId = customer.id
            if comic.isEnable {
                // idを登録
                self.comicId = comic.id
                if comic.currentPurchasePoint.amount <= customer.possessionPoint.amount {
                    // ここまできてようやく正常
                    self.consumptionPoint = comic.currentPurchasePoint
                    self.paymentDateTime = .now
                } else {
                    throw NSError(domain: "所持ポイントが不足しています", code: 0)
                }
            } else {
                throw NSError(domain: "現在取り扱いのできないコミックです", code: 0)
            }
        } else {
            throw NSError(domain: "有効な購入者ではありません", code: 0)
        }
    }
}

// 処理
let currentCustomer = Customer(
    id: Customer.Id(
        id: 1
    ),
    possessionPoint: PurchasePoint(
        amount: 100
    )
)

let selectComic = Comic(
    id: Comic.Id(
        id: 1
    ),
    currentPurchasePoint: PurchasePoint(
        amount: 100
    )
)

do {
    let payment = try PurchasePointPayment(
        customer: currentCustomer,
        comic: selectComic
    )
} catch let e {
    print(e)
}

File: 14/list.14.1.dart
// DartのDateTimeは'timezone'を持つため、'timezone'を無視するためのパッケージを利用します
import 'package:timezone/timezone.dart' as tz;

class Id {
  final int id;
  Id(this.id);
}

class PurchasePoint {
  final int amount;
  PurchasePoint(this.amount);
}

class Customer {
  final Id id;
  final PurchasePoint possessionPoint;

  Customer(this.id, this.possessionPoint);

  // 本来は何かの条件でユーザが有効か無効か決まる
  // 今回は偶数のみ有効
  bool get isEnable => id.id % 2 == 0;
}

class Comic {
  final Id id;
  final PurchasePoint currentPurchasePoint;

  Comic(this.id, this.currentPurchasePoint);

  // 本来は何かの条件でユーザが有効か無効か決まる
  // 今回は偶数のみ有効
  bool get isEnable => id.id % 2 == 0;
}

class PurchasePointPayment {
  final Id customerId;
  final Id comicId;
  final PurchasePoint consumptionPoint;
  final DateTime paymentDateTime;

  PurchasePointPayment._(this.customerId, this.comicId, this.consumptionPoint,
      this.paymentDateTime);

  static PurchasePointPayment create(Customer customer, Comic comic) {
    // [BAD]
    // if elseが多くて入れ子になっているため結局正常な処理がどこか分かりにくい
    if (customer.isEnable) {
      if (comic.isEnable) {
        if (comic.currentPurchasePoint.amount >
            customer.possessionPoint.amount) {
          // ここまできてようやく正常な初期化処理
          return PurchasePointPayment._(customer.id, comic.id,
              comic.currentPurchasePoint, tz.TZDateTime.now(tz.UTC));
        } else {
          throw Exception("所持ポイントが不足しています");
        }
      } else {
        throw Exception("現在取り扱いのできないコミックです");
      }
    } else {
      throw Exception("有効な購入者ではありません");
    }
  }
}

void main() {
  try {
    final currentCustomer = Customer(Id(1), PurchasePoint(100));
    final selectComic = Comic(Id(1), PurchasePoint(100));

    final purchasePointPayment =
        PurchasePointPayment.create(currentCustomer, selectComic);
  } catch (e) {
    print(e);
  }
}
